connect/runtime/src/main/java/org/apache/kafka/connect/runtime/TransformationChain.java

```diff
@@ -0,0 +1,61 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package org.apache.kafka.connect.runtime;
+
+import org.apache.kafka.connect.connector.ConnectRecord;
+import org.apache.kafka.connect.transforms.Transformation;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+public class TransformationChain {
+
+    public static final TransformationChain NO_OP = new TransformationChain(Collections.<Transformation>emptyList());
+
+    private final List<Transformation> transformations;
+
+    public TransformationChain(List<Transformation> transformations) {
+        this.transformations = transformations;
+    }
+
+    @SuppressWarnings("unchecked")
+    public <R extends ConnectRecord<R>> R apply(R record) {
+        if (transformations.isEmpty()) return record;
+
+        for (Transformation transformation : transformations) {
```

Comment:
For generics the type correctness is normally guaranteed generically within the generic class at compile time and casting is validated at the point where the generic type becomes concrete, i.e. at the point where you actually invoke these methods with a known type.

Wouldn't it be better to just push the generic type up to the class level? Put `<R extends ConnectRecord<R>>` on `TransformationChain`, adjust the class to use `R` everywhere, and then in `WorkerSourceTask` and `WorkerSinkTask` you can just instantiate the appropriate typed `TransformationChain<SourceRecord|SinkRecord>`. It seems like this gives you all the type safety you want without the casting and ignoring unchecked exceptions.